Routing tree editor - support new matchers syntax

When testing with the latest example the editor is showing the wrong matched route

https://prometheus.io/docs/alerting/latest/configuration/#example

There is a second issue with this tool.  If you only have a single route (besides the default) the tool errors.

Example config...

```yaml
route:
  receiver: default
  routes:
    - receiver: customer01
      matchers:
      - instance=~".*\\.cus01\\.example\\.com:[0-9]+$|.*customer01\\..*"
      continue: true   # continue so default receiver also gets these notifications
```

gives a bunch of d3js errors in the browser console (`d3.v3.min.js:1 Error: <g> attribute transform: Expected number, "rotate(NaN)translate(45…`) and the rendering of the route is unreadable.

--- a/static/routing-tree.js
+++ b/static/routing-tree.js
@@ -25,6 +25,8 @@ var tooltip = d3.select("body")
     .style("z-index", "10")
     .style("visibility", "hidden");
 
+const promQlRegExp = /\b(?<label>[a-z0-9_]\w*)(?<selector>=~?|![=~])"?(?<value>(?<=")(?:[^\\"]|\\.)*(?=")|\w+[^\s},]+)/gmi;
+
 function parseSearch(searchString) {
   var labels = searchString.replace(/{|}|\"|\'/g, "").split(",");
   var o = {};
@@ -164,6 +166,41 @@ function massage(root, receivers) {
       o.name = key;
       matchers.push(o);
     }
+  }
+
+  if (root.matchers) {
+    root.matchers.forEach((matcher) => {
+      let o = {};
+
+      let matchedExpressions;
+      while ((matchedExpressions = promQlRegExp.exec(matcher)) !== null) {
+        let [match, label, selector, value] = matchedExpressions;
+        o.name = label;
+
+        switch (selector) {
+          case "=~":
+            o.isRegex = true;
+            o.value = new RegExp("^(?:" + value + ")$");
+            matchers.push(o);
+            break;
+          case "!=":
+            o.isRegex = true;
+            o.value = new RegExp("^(?!" + value + "$)");
+            matchers.push(o);
+            break;
+          case "!~":
+            o.isRegex = true;
+            o.value = new RegExp("^(?!" + value + "$)");
+            matchers.push(o);
+            break;
+          case "=":
+            o.isRegex = false;
+            o.value = value
+            matchers.push(o);
+            break;
+        }
+      }
+    });
   }
 
   root.matchers = matchers;

--- a/static/routing-tree.js
+++ b/static/routing-tree.js
@@ -25,7 +25,7 @@ var tooltip = d3.select("body")
     .style("z-index", "10")
     .style("visibility", "hidden");
 
-const promQlRegExp = /\b(?<label>[a-z0-9_]\w*)(?<selector>=~?|![=~])"?(?<value>(?<=")(?:[^\\"]|\\.)*(?=")|\w+[^\s},]+)/gmi;
+const promQlRegExp = /\b(?<label>[a-z0-9_]\w*)\s?(?<selector>=~?|![=~])\s"?(?<value>(?<=")(?:[^\\"]|\\.)*(?=")|\w+[^\s},]+)/gmi;
 
 function parseSearch(searchString) {
   let o = {};

--- a/static/routing-tree.js
+++ b/static/routing-tree.js
@@ -25,7 +25,7 @@ var tooltip = d3.select("body")
     .style("z-index", "10")
     .style("visibility", "hidden");
 
-const promQlRegExp = /\b(?<label>[a-z0-9_]\w*)\s?(?<selector>=~?|![=~])\s"?(?<value>(?<=")(?:[^\\"]|\\.)*(?=")|\w+[^\s},]+)/gmi;
+const promQlRegExp = /\b(?<label>[a-z0-9_]\w*)\s?(?<selector>=~?|![=~])\s?"?(?<value>(?<=")(?:[^\\"]|\\.)*(?=")|\w+[^\s},]+)/gmi;
 
 function parseSearch(searchString) {
   let o = {};
@@ -114,6 +114,13 @@ function matchLabels(matchers, labelSet) {
   return true;
 }
 
+const Matcher = Object.freeze({
+  EQ: "MatchEqual",
+  NE: "MatchNotEqual",
+  RE: "MatchRegexp",
+  NRE: "MatchNotRegexp",
+})
+
 // Compare single matcher to labelSet
 function matchLabel(matcher, labelSet) {
   var v = "";
@@ -121,11 +128,30 @@ function matchLabel(matcher, labelSet) {
     v = labelSet[matcher.name];
   }
 
-  if (matcher.isRegex) {
-    return matcher.value.test(v)
+  if (matcher.op !== undefined) {
+    return matchNewSyntax(matcher, v);
+  } else {
+    // Deprecated matchers check
+    if (matcher.isRegex) {
+      return matcher.value.test(v)
+    }
+    return matcher.value === v;
   }
+}
 
-  return matcher.value === v;
+function matchNewSyntax(matcher, v) {
+  switch (matcher.op) {
+    case Matcher.EQ:
+      return matcher.value === v;
+    case Matcher.NE:
+      return matcher.value !== v;
+    case Matcher.RE:
+    case Matcher.NRE:
+      return matcher.value.test(v);
+    default:
+      console.log("Invalid matcher");
+      break;
+  }
 }
 
 // Load the parsed config and create the tree
@@ -167,33 +193,34 @@ function massage(root, receivers) {
     }
   }
 
+  // PromQL matcher syntax check
   if (root.matchers) {
     root.matchers.forEach((matcher) => {
       let o = {};
-
       let matchedExpressions;
+
       while ((matchedExpressions = promQlRegExp.exec(matcher)) !== null) {
         let [match, label, selector, value] = matchedExpressions;
         o.name = label;
 
         switch (selector) {
           case "=~":
-            o.isRegex = true;
             o.value = new RegExp("^(?:" + value + ")$");
+            o.op = Matcher.RE;
             matchers.push(o);
             break;
           case "!=":
-            o.isRegex = true;
-            o.value = new RegExp("^(?!" + value + "$)");
+            o.op = Matcher.NE
+            o.value = value;
             matchers.push(o);
             break;
           case "!~":
-            o.isRegex = true;
+            o.op = Matcher.NRE
             o.value = new RegExp("^(?!" + value + "$)");
             matchers.push(o);
             break;
           case "=":
-            o.isRegex = false;
+            o.op = Matcher.EQ
             o.value = value
             matchers.push(o);
             break;

--- a/static/routing-tree.js
+++ b/static/routing-tree.js
@@ -28,12 +28,11 @@ var tooltip = d3.select("body")
 const promQlRegExp = /\b(?<label>[a-z0-9_]\w*)(?<selector>=~?|![=~])"?(?<value>(?<=")(?:[^\\"]|\\.)*(?=")|\w+[^\s},]+)/gmi;
 
 function parseSearch(searchString) {
-  var labels = searchString.replace(/{|}|\"|\'/g, "").split(",");
-  var o = {};
-  labels.forEach(function(label) {
-    var l = label.trim().split("=");
-    o[l[0]] = l[1];
-  });
+  let o = {};
+  let matchedExpressions;
+  while ((matchedExpressions = promQlRegExp.exec(searchString)) !== null) {
+    o[matchedExpressions[1]] = matchedExpressions[3];
+  }
   return o;
 }
 

